fsw/src/generic_fss_device.c

```diff
@@ -0,0 +1,189 @@
+    {
+        #ifdef GENERIC_FSS_CFG_DEBUG
+            OS_printf("  GENERIC_FSS_RequestData: GENERIC_FSS_CommandDevice reported error %d \n", status);
+        #endif 
+    }
+    return status;
+}
+
+/* 
+** Generic read data from device
+*/
+static int32_t GENERIC_FSS_ReadData(spi_info_t *device, uint8_t* write_data, uint8_t* read_data, uint8_t data_length)
+{
+    int32_t status = OS_ERROR;
+    int32_t bytes = 0;
+    int32_t bytes_available = 0;
+    uint8_t ms_timeout_counter = 0;
+
+    /* Read data */
+    if ((spi_select_chip(device) == SPI_SUCCESS) &&
+        (spi_transaction(device, write_data, read_data, data_length, GENERIC_FSS_CFG_DELAY, GENERIC_FSS_CFG_BITS_PER_WORD, 0) == SPI_SUCCESS) &&
+        (spi_unselect_chip(device) == SPI_SUCCESS)) {
+        status = OS_SUCCESS;
+    }
+
+    return status;
+}
+
+
+/* 
+** Generic command to device
+** Note that confirming the echoed response is specific to this implementation
+*/
+static int32_t GENERIC_FSS_CommandDevice(spi_info_t *device, uint8_t cmd_code)
+{
+    int32_t status = OS_ERROR;
+    int32_t bytes = 0;
+    uint8_t write_data[GENERIC_FSS_DEVICE_CMD_SIZE];
+    uint8_t read_data[GENERIC_FSS_DEVICE_CMD_SIZE];
+
+    /* Prepare command */
+    write_data[0] = GENERIC_FSS_DEVICE_HDR_0;
+    write_data[1] = GENERIC_FSS_DEVICE_HDR_1;
+    write_data[2] = GENERIC_FSS_DEVICE_HDR_2;
+    write_data[3] = GENERIC_FSS_DEVICE_HDR_3;
+    write_data[4] = cmd_code;
+    write_data[5] = 0x01;
+    write_data[6] = compute_checksum(write_data, 4, 2);
+    #ifdef GENERIC_FSS_CFG_DEBUG
+        OS_printf("  GENERIC_FSS_CommandDevice[%d] = ", status);
+        for (uint32_t i = 0; i < GENERIC_FSS_DEVICE_CMD_SIZE; i++)
+        {
+            OS_printf("%02x", write_data[i]);
+        }
+        OS_printf("\n");
+    #endif
+
+    /* Write data */
+    if ((spi_select_chip(device) == SPI_SUCCESS) &&
+        (spi_transaction(device, write_data, read_data, GENERIC_FSS_DEVICE_CMD_SIZE, GENERIC_FSS_CFG_DELAY, GENERIC_FSS_CFG_BITS_PER_WORD, 0) == SPI_SUCCESS) &&
+        (spi_unselect_chip(device) == SPI_SUCCESS)) {
+        status = OS_SUCCESS;
+    }
+    
+    return status;
+}
+
+static uint8_t compute_checksum(uint8_t in[], int starting_byte, int number_of_bytes)
+{
+    uint32_t sum = 0;
+    uint8_t checksum;
+    for (int i = starting_byte; i < starting_byte + number_of_bytes; i++) {
+        sum += in[i];
+    }
+    checksum = (uint8_t)(sum & 0x000000FF);
+    return checksum;
+}
+
+static float fourbytes_little_endian_to_float(uint8_t bytes[])
+{
+    union {float f; uint32_t u;} fu;
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    fu.u = ((uint32_t)bytes[3] << 24) | ((uint32_t)bytes[2] << 16) | ((uint32_t)bytes[1] << 8) | (uint32_t)bytes[0];
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+    fu.u = ((uint32_t)bytes[0] << 24) | ((uint32_t)bytes[1] << 16) | ((uint32_t)bytes[2] << 8) | (uint32_t)bytes[3];
+#else
+    #error "__BYTE_ORDER__ is not defined"
+#endif
+    return fu.f;
+}
```

Comment:
Should we move standard functions like this to hwlib for reuse?
https://github.com/nasa-itc/hwlib/blob/dev/fsw/public_inc/hwlib.h#L47

Reply:
Maybe.  It might be a good idea to have some more experienced (with endianness issues) folks comment on how bullet proof this solution is.  I think it is pretty good, but if it is going into a library for wider use I wouldn't mind some extra review.